receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpuscraper
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/host"
+	"go.opencensus.io/trace"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdata"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/pdatautil"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/receiver/hostmetricsreceiver/internal"
+)
+
+// Scraper for CPU Metrics
+type Scraper struct {
+	config    *Config
+	consumer  consumer.MetricsConsumer
+	startTime pdata.TimestampUnixNano
+}
+
+// NewCPUScraper creates a set of CPU related metrics
+func NewCPUScraper(ctx context.Context, cfg *Config, consumer consumer.MetricsConsumer) (*Scraper, error) {
+	return &Scraper{config: cfg, consumer: consumer}, nil
+}
+
+// Start
+func (c *Scraper) Start(ctx context.Context) error {
+	bootTime, err := host.BootTime()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		c.startTime = pdata.TimestampUnixNano(bootTime)
+
+		ticker := time.NewTicker(c.config.CollectionInterval())
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				c.scrapeMetrics(ctx)
+			case <-ctx.Done():
```

Comment:
This is not how we intend to close the pipeline. It is not guaranteed that we mark the start Context as done before we end. Can you please describe how you expect this Scraper to be closed?

Reply:
I was loosely following the pattern Jay referenced here: https://github.com/open-telemetry/opentelemetry-collector/pull/848#discussion_r411730113 (see my comment below), and per this example: [k8sclusterreceiver example](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8e3e28d3016de4e1da618d3001245e4faeb008db/receiver/k8sclusterreceiver/receiver.go#L39), but given that the scrapers run within the context of the host metrics receiver, I extended it to so that the parent context (that is in the scope of the hostmetrics receiver) controls when the context is cancelled.

i.e. the hostmetrics receiver creates a new `context.WithCancel(ctx)` in its `Start` function (line 63) and calls the returned context's `cancel` function in its `Shutdown` function (line 78). The Scrapers are always started in the context of the host metrics receiver so are guaranteed to receive this signal.

Maybe creating that dependency between the parent context is not considered great practice in general? (although that seems to be what its inteded for). Regardless, it definitely adds confusion (and could potentially cause some kind of race condition) to both: a) listen to the parent context `Done()` channel, and b) have a `Close()` function that the parent is expected to call.

I've removed the creation of the cancelable context from host metrics receiver, and moved this down to the cpu scraper itself which uses `ctx.Done()` to signal the termination of the ticker (I could change it to just using a regular channel, but figure its basically the same)